loader: FrozenImporter: handle encodings when reading source files

rokm opened the pull request:
When reading collected .py source files via `FrozenImporter`'s `get_source()` method, open the file with encoding explicitly set to UTF-8 (the default encoding for source files in Python 3). Failing to specify encoding leads to platform-specific default being used (e.g., `cp1252` on Windows), which results in an error if the source file contains unicode emojis.

Fixes #6143.

rokm committed:
loader: FrozenImporter: handle encodings when reading source files

When reading collected .py source files via FrozenImporter's
get_source() method, detect the source file's encoding and
properly decode it. The decoding function is based on the
CPython's implementation of the same functionality, which handles
both encodings and newline normalization.

Fixes #6143.

rokm committed:
tests: add a test for retrieving source code that contains utf-8 emojis

Add a test that ends up using FrozenImporter.get_source() to load
a source code file that contains utf-8 emojis. Reproduces scenario
from issue #6143.

bwoodsend commented:
As I understand it, that would break for non UTF8 files? Looking at [importlib's get_source() method](https://github.com/python/cpython/blob/4ceec495598e78f0776dd46d511dcc612a434dc3/Lib/importlib/_bootstrap_external.py#L943-L951), the cpython implementation reads the source file in binary mode, then [uses tokenize.detect_encoding() to decode it](https://github.com/python/cpython/blob/4ceec495598e78f0776dd46d511dcc612a434dc3/Lib/importlib/_bootstrap_external.py#L715), then normalises line endings using some hidden corner of `io` (we'd probably be better just to use `source.replace("\r\n", "\n").replace("\r", "\n")` than use that hidden corner).

rokm commented:
Let's try using (a copy of) CPython's implementation, then, and properly handle all encodings. The `get_source()` will probably not be used during bootstrap, so we can locally import `io` (which also provides that `IncrementalNewlineDecoder`) and `tokenize`. I suppose the question is, do we need to take any extra steps to ensure that `tokenize` is collected (`io` already seems to be a part of the base library zip)?

bwoodsend commented:
We could assume the worst and add `tokenize` to `PyInstaller.compat.PY3_BASE_MODULES`.

rokm committed:
compat: add tokenize to PY3_BASE_MODULES

This way, we can ensure that it is always available for use in
the _decode_source() function in loader/pyimod03_importers.py

s0undt3ch commented:
When can we expect this change in a release?

bwoodsend commented:
Looks like we forgot to backport this. I've done it now so it'll be in the next release (but I have no idea when that'll be).

s0undt3ch commented:
How about a 4.7.1 release?

bwoodsend commented:
A 4.7.1 release is actually more work for us than a 4.8.0 release because we'd then need to create a new branch off this one and drop all the feature commits.

s0undt3ch commented:
Bummer.
Ok, I'll watch out for the next release.
